### PR TITLE
fix stale writes by locking collections when auth/connection issue discovered

### PR DESCRIPTION
**Pull request by twrichards**
https://trello.com/c/B3dnmNti/2094-its-possible-to-publish-a-stale-collection-in-the-fronts-tool

## What's changed?
As per https://docs.google.com/document/d/14vbdereyeYU8MW6VDdUA_Pbbnv8_tf9ZJciKZMqpd0w/edit# it was possible for stale writes to occur in fronts, due to a combination of the 10second polling and people having auth/connection issues.

This PR seeks to address that by locking the collections when the latest fetch for collections is either errored or is stale by 20 seconds. By locking only the collections we allow users to potentially continue other productive work in the tool such as racking items up in the clipboard.

Before being able to do this there was a bit of re-work required, most notably...

- `FETCH_SUCCESS` events were only fired for collections when the data actually changed, and as such the `lastFetch` and `error` fields were not updated until the next material change to the collections (made by someone else), as such without changing, this the new locking feature would have displayed for much longer than necessary.

- Previously in `pandaFetch` any network failures would've thrown errors which were not caught in the [somewhat convoluted] async/await/Promise construct, and came out like this (Offline is forced on here, notice the `Uncaught` line)....
image
... they're now caught and calls the `reject` of the outer promise (this in turn populates the `error` field under `collections` in the state).

## GIF
fronts_stale_fix

### General
- [ ] 🤖 Relevant tests added
- [x] ✅ CI checks / tests run locally
- [x] 🔍 Checked on CODE

### Client
- [x] 🚫 No obvious console errors on the client (i.e. React dev mode errors)
- [x] 🎛️ No regressions with existing user interactions (i.e. all existing buttons, inputs etc. work)
- [x] 📷 Screenshots / GIFs of relevant UI changes included

### CODE REVIEW

File: fronts-client/src/lib/createAsyncResourceBundle/index.ts
Comment (twrichards) [inline fronts-client/src/lib/createAsyncResourceBundle/index.ts:394]: based on the `if` above this appeared redundant to me

File: fronts-client/src/services/pandaFetch.ts
Comment (twrichards) [inline fronts-client/src/services/pandaFetch.ts:57]: without this new outer `try`/`catch` we would get badness in the console...

File: fronts-client/src/services/pandaFetch.ts
Comment (twrichards) [inline fronts-client/src/services/pandaFetch.ts:57]: without this new outer `try`/`catch` we would get badness in the console...

File: fronts-client/src/components/FrontsEdit/FrontContent.tsx
Comment (twrichards) [inline fronts-client/src/components/FrontsEdit/FrontContent.tsx:170]: Whilst I'm not sure this is actually necessary in that we can check the error directly, I doubt it actually works as desired, as the only time its evaluated is when some thing causes it to re-render (which is likely fresh changes - at which point the lastFetch will be v. recent)... instead perhaps we use a `setTimeout` (with 20 second delay) to write some new field to state (something like `lastStaleCheck`) to force this to be re-evaluated (there might be a nice React way of doing that)... thoughts welcome....

Comment (twrichards) [reply fronts-client/src/components/FrontsEdit/FrontContent.tsx:183]: DONE in https://github.com/guardian/facia-tool/pull/1277/commits/da957b710fb4802b478d304481817b5acac588d4

File: fronts-client/src/components/FrontsEdit/FrontContent.tsx
Comment (jonathonherbert) [inline fronts-client/src/components/FrontsEdit/FrontContent.tsx:23]: TIL, numeric separators! Handy!

File: fronts-client/src/actions/Collections.ts
Comment (jonathonherbert) [inline fronts-client/src/actions/Collections.ts:247]: Might be worth leaving a comment here to indicate why we're dispatching this. Other devs may consider it and remove, introducing a bug – and there's no test to catch it AFAICS (not necessarily saying add a test, it may be time consuming!)

File: fronts-client/src/services/pandaFetch.ts
Comment (jonathonherbert) [inline fronts-client/src/services/pandaFetch.ts:54]: This change is a Good Thing, but we've no way at present to assess its impact on users.

I'd recommend capturing this failure in Raven (it's got retry strats, this isn't as paradoxical as it sounds 😁 ) so we can see if users are experiencing high volumes of these messages, and diagnose as appropriate.

There's an example in `move.ts`, e.g. `Raven.captureMessage(`Moving item up in clipboard failed: ${e.message}`);` (but let's not chuck away the stack trace!)

In any case, great catch. Pun not intended.

File: fronts-client/src/components/FrontsEdit/FrontContent.tsx
Comment (jonathonherbert) [inline fronts-client/src/components/FrontsEdit/FrontContent.tsx:252]: We shouldn't access redux state directly. Instead, there are selectors in `asyncResourceBundle` for both of these pieces of state.

**Comment by jonathonherbert** [comment]
This is great. Left a few recommendations.

I'm a little uncomfortable signposting in the overlay to a notification that the user can dismiss out of hand. Is there a reason we can't include the reason that the collection is locked in the overlay itself? E.g. `We couldn't refresh this collection. It may be out of date.` – but! don't use this messaging! Talk to CP 😁

File: fronts-client/src/services/pandaFetch.ts
Comment (twrichards) [reply fronts-client/src/services/pandaFetch.ts:54]: happy to add the call to Sentry, are you recommending that instead of a user banner?

File: fronts-client/src/services/pandaFetch.ts
Comment (jonathonherbert) [reply fronts-client/src/services/pandaFetch.ts:58]: I don't disagree – I just think it's important that we understand the impact.

File: fronts-client/src/actions/Collections.ts
Comment (twrichards) [reply fronts-client/src/actions/Collections.ts:249]: added a comment for now, will look into adding a test too

File: fronts-client/src/components/FrontsEdit/FrontContent.tsx
Comment (twrichards) [reply fronts-client/src/components/FrontsEdit/FrontContent.tsx]: Then DONE better in https://github.com/guardian/facia-tool/pull/1277/commits/3e0fa48b3f54b82e4f4a4d082cab76d799ccac1b

File: fronts-client/src/services/pandaFetch.ts
Comment (twrichards) [reply fronts-client/src/services/pandaFetch.ts:62]: added in addition to the banners (as discussed over chat) - see https://github.com/guardian/facia-tool/pull/1277/commits/5884f19903acdf639dff6849012a242567f8e32d

**Review by jonathonherbert — APPROVED** [review]
Tested locally, works as expected – any connection errors lock collection editing, latent connections have a tolerance window of 30s before we lock collection editing.

Minor: recommend a little margin and possible centre-justifying text to match header, as the text abuts the collection bounds on smaller screen.

Screenshot 2020-12-07 at 09 54 47

This PR goes a little further than I originally thought by locking collections immediately when there are errors, rather than waiting for collections to be stale regardless of network conditions. I'm in two minds about this, as we don't know how often connection problems occur, and whether for users on poor connections we're going to get e.g. intermittent flickers etc.

Let's pay attention to what Sentry picks up to understand how this affects users. I'd be tempted to display connection errors but avoid locking collections immediately, instead waiting for them to become stale before preventing edits – but I'm happy to approve as long as we've got a way of measuring the impact, if only by proxy.

**Comment by twrichards** [comment]
> Minor: recommend a little margin and possible centre-justifying text to match header, as the text abuts the collection bounds on smaller screen.
> 
> Screenshot 2020-12-07 at 09 54 47

Good spot, thanks @jonathonherbert (wide monitor catching me out again). Done in https://github.com/guardian/facia-tool/pull/1277/commits/d108e9336cefdecbeac27b78f535b031911f4b34
image

**Pull request merged**
